nomos-libp2p/src/behaviour/nat/address_mapper/nat_pmp.rs

```diff
@@ -0,0 +1,158 @@
+            .read_response_or_retry()
+            .await
+            .map_err(|e| AddressMapperError::PortMappingFailed(e.to_string()))?
+        {
+            Response::UDP(r) | Response::TCP(r) => Ok(r.public_port()),
+            Response::Gateway(_) => Err(AddressMapperError::PortMappingFailed(
+                "Expected UDP/TCP mapping response; got Gateway response".to_owned(),
+            )),
+        }
+    }
+
+    async fn query_public_ip(&mut self) -> Result<Ipv4Addr, AddressMapperError> {
+        self.nat_pmp
+            .send_public_address_request()
+            .await
+            .map_err(|e| AddressMapperError::GatewayDiscoveryFailed(e.to_string()))?;
+
+        match self
+            .nat_pmp
+            .read_response_or_retry()
+            .await
+            .map_err(|e| AddressMapperError::GatewayDiscoveryFailed(e.to_string()))?
+        {
+            Response::Gateway(pa) => Ok(*pa.public_address()),
+            other => Err(AddressMapperError::GatewayDiscoveryFailed(format!(
+                "Expected PublicAddress response; got {other:?}"
+            ))),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl MappingProtocol for NatPmp {
+    async fn initialize(settings: NatMappingSettings) -> Result<Box<Self>, AddressMapperError>
+    where
+        Self: Sized,
+    {
+        let nat_pmp = new_tokio_natpmp()
+            .await
+            .map_err(|e| AddressMapperError::PortMappingFailed(e.to_string()))?;
+
+        Ok(Box::new(Self { settings, nat_pmp }))
+    }
+
+    async fn map_address(&mut self, address: &Multiaddr) -> Result<Multiaddr, AddressMapperError> {
+        let (port, protocol) = extract_port_and_protocol(address)?;
+
+        self.send_map_request(protocol, port).await?;
+
+        let public_port = timeout(self.settings.timeout, self.recv_map_response_public_port())
+            .await
+            .map_err(|_| {
+                AddressMapperError::PortMappingFailed(
+                    "Timeout waiting for NAT-PMP mapping response".to_owned(),
+                )
+            })??;
+
+        let public_ip = timeout(self.settings.timeout, self.query_public_ip())
+            .await
+            .map_err(|_| {
+                AddressMapperError::PortMappingFailed(
+                    "Timeout waiting for NAT-PMP public IP response".to_owned(),
+                )
+            })??;
+
+        public_address(address, public_ip, public_port)
+    }
+}
+
+fn extract_port_and_protocol(addr: &Multiaddr) -> Result<(u16, Protocol), AddressMapperError> {
+    addr.iter()
+        .find_map(|p| match p {
+            MaProto::Tcp(p) => Some((p, Protocol::TCP)),
+            MaProto::Udp(p) => Some((p, Protocol::UDP)),
+            _ => None,
+        })
+        .ok_or_else(|| {
+            AddressMapperError::MultiaddrParseError(
+                "No TCP or UDP port found in multiaddr".to_owned(),
+            )
+        })
+}
+
+pub fn public_address(
+    internal: &Multiaddr,
```

Comment:
Should we take a `Multiaddr` instead of a reference to give a hint the old one is not meant to be used anymore? Although `Multiaddr` implements `Copy`, I think it would still be a better design.

Reply:
Or maybe call this function `to_public`.